core/arch/riscv64/mangle.c

```diff
@@ -385,13 +394,406 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+                         instr_t *next_instr)
+{
+    if (!instr_uses_reg(instr, dr_reg_stolen) && !instr_uses_reg(instr, DR_REG_TP))
+        return next_instr;
+
+    if (instr_is_cbr(instr)) {
+        mangle_cbr_stolen_reg_and_tp_reg(dcontext, ilist, instr, instr_get_next(instr));
+    } else if (!instr_is_mbr(instr)) {
+        mangle_stolen_reg_and_tp_reg(dcontext, ilist, instr, next_instr);
+    }
+
+    return next_instr;
+}
+
+static instr_t *
+mangle_exclusive_load(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+                      instr_t *next_instr)
+{
+    /* TODO i#3544: Not implemented.  */
+    ASSERT_NOT_IMPLEMENTED(!instr_uses_reg(instr, DR_REG_TP));
+
+    reg_id_t scratch_reg1 = DR_REG_NULL, scratch_reg2 = DR_REG_NULL;
+    ushort slot1, slot2;
+    int aqrl, opcode;
+    opnd_t dst, src0;
+    opnd_size_t opsz;
+    bool uses_reg_stolen;
+    ASSERT(instr_is_exclusive_load(instr));
+    ASSERT(instr_num_dsts(instr) == 1 && instr_num_srcs(instr) == 2 &&
+           opnd_is_immed_int(instr_get_src(instr, 1)));
+
+    aqrl = opnd_get_immed_int(instr_get_src(instr, 1));
+    uses_reg_stolen = instr_uses_reg(instr, dr_reg_stolen);
+
+    /* Pick and spill scratch register(s). */
+    scratch_reg1 = pick_scratch_reg(dcontext, instr, DR_REG_NULL, &slot1);
+    PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg1, slot1));
+
+    if (uses_reg_stolen) {
+        scratch_reg2 = pick_scratch_reg(dcontext, instr, scratch_reg1, &slot2);
+        PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg2, slot2));
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch_reg2, TLS_REG_STOLEN_SLOT));
+    }
+
+    /* Keep the release semantics if needed. */
+    if (aqrl & LRSC_ORDERING_RL_MASK) {
+        /* fence rw, rw */
+        PRE(ilist, instr,
+            INSTR_CREATE_fence(dcontext,
+                               opnd_create_immed_int(FENCE_ORDERING_RW, OPSZ_4b),
+                               opnd_create_immed_int(FENCE_ORDERING_RW, OPSZ_4b),
+                               opnd_create_immed_int(FENCE_FM_NONE, OPSZ_4b)));
+    }
+
+    /* Replace exclusive load to a normal one. */
+    dst = instr_get_dst(instr, 0);
+    src0 = instr_get_src(instr, 0);
+    opcode = instr_get_opcode(instr) == OP_lr_d ? OP_ld : OP_lw;
+    opsz = opcode == OP_ld ? OPSZ_8 : OPSZ_4;
+    ASSERT(opnd_is_reg(dst) && opnd_is_reg(src0));
+    if (opnd_get_reg(dst) == dr_reg_stolen) {
+        opnd_replace_reg(&dst, dr_reg_stolen, scratch_reg2);
+    }
+    if (opnd_get_reg(src0) == dr_reg_stolen) {
+        opnd_replace_reg(&src0, dr_reg_stolen, scratch_reg2);
+    }
+    instr_reset(dcontext, instr);
+    instr_set_opcode(instr, opcode);
+    instr_set_num_opnds(dcontext, instr, 1, 1);
+    instr_set_dst(instr, 0, dst);
+    instr_set_src(
+        instr, 0,
+        opnd_add_flags(opnd_create_base_disp(opnd_get_reg(src0), DR_REG_NULL, 0, 0, opsz),
```

Comment:
You call opnd_create_base_disp above in this file but don't add this printing flag there.  Shouldn't this be auto-added for all base-disp for RISC-V?